### PR TITLE
[FlatList] Add ListEmptyComponent prop

### PR DESCRIPTION
**Pull request by Minishlink**
Hey there :)

Please let me know if the name `ListEmptyComponent` should be changed. I also thought about `ListNoItemsComponent`. Or maybe `ListPlaceholderComponent`?

- [x] Explain the **motivation** for making this change.
- [x] Provide a **test plan** demonstrating that the code is solid.
- [x] Match the **code formatting** of the rest of the codebase.
- [x] Target the `master` branch, NOT a "stable" branch.

## Motivation (required)

In a FlatList, I wanted to show some placeholder when my data is empty (while keeping eventual Header/Footer/RefreshControl).
A way around this issue would be to do something like adding a `ListHeaderComponent` that checks if the list is empty, like so:
```js
ListHeaderComponent={() => (!data.length ? <Text style={styles.noDataText}>No data found</Text> : null)}
```
But I felt it was not easily readable as soon as you have an actual header.

This PR adds a `ListEmptyComponent` that is rendered when the list is empty. 

## Test Plan (required)

I added tests for VirtualizedList, FlatList and SectionList and ran `yarn test -- -u`. I then checked that the snapshots changed like I wanted.
I also tested this against one of my project, though I had to manually add my changes because the project is on RN 0.43.

Here are the docs screenshots:
- [VirtualizedList](https://cloud.githubusercontent.com/assets/82368/25566000/0ebf2b82-2dd2-11e7-8b80-d8c505f1f2d6.png)
- [FlatList](https://cloud.githubusercontent.com/assets/82368/25566005/2842ab42-2dd2-11e7-81b4-32c74c2b4fc3.png)
- [SectionList](https://cloud.githubusercontent.com/assets/82368/25566010/368aec1e-2dd2-11e7-9425-3bb5e5803513.png)

## Next Steps

Thanks for your work!

### CODE REVIEW

**Comment by Minishlink** [comment]
Flow apparently fails to treat the type `?(ReactClass<any> | React.Element<any>)`. I suspect that's why these props (ListHeaderComponent etc) were not put in this file but in the Flat|SectionList files. I feel like it's better to make the props appear on the docs and suppress those errors. I'm not a Flow pro so if somebody has some insight on how to fix the errors, I'd be grateful for some help.

File: Libraries/Lists/__tests__/SectionList-test.js [hunk 1]
```diff
@@ -26,6 +26,16 @@ describe('SectionList', () => {
     );
     expect(component).toMatchSnapshot();
   });
+  it('renders empty list with empty component', () => {
```
Comment (sahrens) [inline Libraries/Lists/__tests__/SectionList-test.js]: I wouldn't worry about duplicating these empty cases in FlatList and SectionList - just VirtualizedList should be fine. Adding to the "bells and whistles" cases is good though.

File: Libraries/Lists/__tests__/VirtualizedList-test.js [hunk 2]
```diff
@@ -48,10 +48,22 @@ describe('VirtualizedList', () => {
     expect(component).toMatchSnapshot();
   });
 
+  it('renders empty list with empty component', () => {
+    const component = ReactTestRenderer.create(
+      <VirtualizedList
+        data={[]}
+        ListEmptyComponent={() => <empty />}
```
Comment (sahrens) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: Should also test the rendered element case.

File: Libraries/Lists/VirtualizedList.js [hunk 3]
```diff
@@ -83,6 +83,21 @@ type OptionalProps = {
...
   /**
+   * Rendered when the list is empty. Can be a React Component Class, a render function, or
+   * a rendered element.
+   */
+  ListEmptyComponent?: ?(ReactClass<any> | React.Element<any>),
+  /**
+   * Rendered at the bottom of all the items. Can be a React Component Class, a render function, or
+   * a rendered element.
+   */
+  ListFooterComponent?: ?(ReactClass<any> | React.Element<any>),
```
Comment (sahrens) [inline Libraries/Lists/VirtualizedList.js]: Thanks for adding these in - it was just an oversight that I left them out.

File: Libraries/Lists/__tests__/VirtualizedList-test.js (hunk 2 above)
Comment (sahrens) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: Can you add headers and footers here too? I want to make sure we test the logic of whether we show them in the empty case.

**Comment by sahrens** [comment]
@gabelevi: thoughts on the flow issue? Does flow understand type refinement with `React.isValidElement`?

File: Libraries/Lists/__tests__/VirtualizedList-test.js (hunk 2 above)
Comment (Minishlink) [inline Libraries/Lists/__tests__/VirtualizedList-test.js]: Sure, done :)

**Comment by Minishlink** [comment]
> What is your thinking around rendering the headers and footers in the empty case? I think you're probably doing the right thing here, but worth giving some extra thought.

Yep, the reason behind that is that if you have a header/footer, you want to have a consistent experience and don't lose the user in the UI when there are no items. After all, if you put a header, it's because it's useful to the user. That's also why I came up with the "Placeholder" name. Maybe this name would be better after all?

File: Libraries/Lists/VirtualizedList.js (hunk 3 above)
Comment (andres-torres-marroquin) [inline Libraries/Lists/VirtualizedList.js]: I think ListFooterComponent doesn't accept React.Element, just ReactClass. 
I got this when I tried ListFooterComponent={<OneComponent/>}
